lib/Auth/Process/SwitchAuth.php

```diff
@@ -117,20 +119,37 @@ public function process(&$state)
                 && !AuthnContextHelper::MFAin([$upstreamContext])
         ); // switch to MFA if preferred and not already done if we handle the proxy mode
 
+        $assurance = '';
+        if (in_array(AuthSwitcher::SFA, $usersCapabilities, true) && in_array(
+            AuthSwitcher::MFA,
+            $usersCapabilities,
+            true
+        )) {
+            $assurance = AuthSwitcher::MFA;
+        } elseif (1 === count($usersCapabilities)) {
+            $assurance = $usersCapabilities[0];
+        }
+        $state['Attributes'][self::ASSURANCE] = [];
+
         if ($performMFA) {
             // MFA
-            $this->performMFA($state);
+            $this->performMFA($state, $assurance);
         } elseif (empty($upstreamContext)) {
-            // SFA
-            self::setAuthnContext($state);
+            self::setAuthnContext($state, $assurance);
         }
     }
 
-    public static function setAuthnContext(&$state)
+    public static function setAuthnContext(&$state, $assurance)
     {
-        $possibleReplies = Utils::wasMFAPerformed(
-            $state
-        ) ? AuthSwitcher::REPLY_CONTEXTS_MFA : AuthSwitcher::REPLY_CONTEXTS_SFA;
+        $mfa_performed = Utils::wasMFAPerformed($state);
+
+        if (AuthSwitcher::SFA === $assurance) {
+            $state['Attributes'][self::ASSURANCE][] = AuthSwitcher::MAX_AUTH;
+        } elseif (AuthSwitcher::MFA === $assurance && $mfa_performed) {
+            $state['Attributes'][self::ASSURANCE][] = AuthSwitcher::MAX_AUTH;
+        }
+
+        $possibleReplies = $mfa_performed ? AuthSwitcher::REPLY_CONTEXTS_MFA : AuthSwitcher::REPLY_CONTEXTS_SFA;
```

Comment:
```suggestion
        $mfaPerformed = Utils::wasMFAPerformed($state);

        if (AuthSwitcher::SFA === $assurance) {
            $state['Attributes'][self::ASSURANCE][] = AuthSwitcher::MAX_AUTH;
        } elseif (AuthSwitcher::MFA === $assurance && $mfaPerformed) {
            $state['Attributes'][self::ASSURANCE][] = AuthSwitcher::MAX_AUTH;
        }

        $possibleReplies = $mfaPerformed ? AuthSwitcher::REPLY_CONTEXTS_MFA : AuthSwitcher::REPLY_CONTEXTS_SFA;
```
everything around is camelCase